### DISCUSSION TITLE
Command to restore from backup

### DISCUSSION THREAD

**Discussion by djotanov** [apache/kvrocks#483] [General]
Are there any plans to implement a new command for restoring from last backup?
I know that redis does not have this command, but it would be very useful for kvrocks to have it (something like RESTORE_LAST_BACKUP).

**Comment by git-hulk** [comment]
Yes, maybe we can add a new flag like `--restore-from-backup` to make the backup restore easier. How do you think? @ShooterIT

**Reply by djotanov** [reply]
You mean start up flag? That would work as well. 
I thought more like a custom command that could be sent from client. The code already exist (restoring backup for replicas), it would just need to be exposed for master instances through a command.

**Reply by git-hulk** [reply]
Yes, what I mean was the start-up flag, I think it should be dangerous to restore the backup online.

**Reply by djotanov** [reply]
Thanks,

I read a bit through the source code, and saw that bgsave actually creates a checkpoint, not a real backup. So that means I can't use remote file system for backups right, since checkpoints create hard links?

In that case, would it work if I create a cron job which rsyncs backup folder to remote file system. And when I need to restore the master, can I just copy the files (checkpoint) from backup folder from remote file system into DB folder (previously deleting the old DB folder). Would that be sufficient or I would need to do something else to the checkpoint?

**Reply by git-hulk** [reply]
Yes, u are right that backup in Kvrocks is checkpoint instead of backup. For syncing to the remote filesystem, it is sufficient to use rsync before deleting the backup folder. Also the restore is as simple as to replace the db dir.

**Reply by djotanov** [reply]
Thanks

**Reply by ShooterIT** [reply]
AFAIK, if the checkpoint dir is remote file system, Rocksdb also copy files instead of hard link.

**Comment by ShooterIT** [comment]
> Are there any plans to implement a new command for restoring from last backup?

It looks like a simple Data Rollback. Could you describe your scene？

**Reply by djotanov** [reply]
I'm deploying kvrocks to kubernetes. Even though restore is just replacing db folder with backup folder, it still requires a bit more work to implement recovery in kubernetes. It is harder to automate. I would need to write a separate process which restarts stateful set, and stateful set should have init container which would do the restoration prior to kvrocks starting.
If there were a restore kvrocks command, this would be as simple as sending a command from connected client.

**Reply by Vseg1** [reply]
Hello @djotanov, I am too deploying kvrocks on a Kubernetes infrastructure and I am unable to backup and restore the database using the data folder. Do you have any procedure or tips regarding how to backup consistently the entirety of the database on a remote storage such as a bucket S3? I guess I am missing something regarding the backup workflow on kvrocks.
